File: crates/core/src/harness/caps.rs
use serde::{Deserialize, Serialize};
use specta::Type;

use crate::model::{HarnessId, ItemKind};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct OpSupport {
    pub project: bool,
    pub global: bool,
}

pub const BOTH: OpSupport = OpSupport {
    project: true,
    global: true,
};
pub const PROJECT: OpSupport = OpSupport {
    project: true,
    global: false,
};
pub const GLOBAL: OpSupport = OpSupport {
    project: false,
    global: true,
};
pub const NONE: OpSupport = OpSupport {
    project: false,
    global: false,
};

/// What each operation supports for one harness × kind. `observe` is derived
/// from adapter surface declarations (tested); mutation columns land with
/// their phases and stay honest through those phases' tests.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct KindCaps {
    pub observe: OpSupport,
    pub adopt: OpSupport,
    pub install: OpSupport,
    pub toggle: OpSupport,
    pub remove: OpSupport,
    pub refresh: OpSupport,
}

const fn unsupported() -> KindCaps {
    KindCaps {
        observe: NONE,
        adopt: NONE,
        install: NONE,
        toggle: NONE,
        remove: NONE,
        refresh: NONE,
    }
}

/// Fully managed at the given scopes.
const fn managed(scopes: OpSupport) -> KindCaps {
    KindCaps {
        observe: scopes,
        adopt: scopes,
        install: scopes,
        toggle: scopes,
        remove: scopes,
        refresh: scopes,
    }
}

/// Read-only surface: scanning works, nothing may be written.
const fn observe_only(scopes: OpSupport) -> KindCaps {
    KindCaps {
        observe: scopes,
        adopt: NONE,
        install: NONE,
        toggle: NONE,
        remove: NONE,
        refresh: NONE,
    }
}

/// Format facts per harness — owned here beside the op table so renderers
/// and the surface model read one source of truth instead of scattering
/// literals. Extended axis by axis as consumers land.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct FormatCaps {
    /// Bytes the harness reliably loads from one SKILL.md before it
    /// truncates; `None` means no known cap. Oversized bodies split into
    /// `references/` rather than truncating.
    pub skill_body_max_bytes: Option<usize>,
}

pub const fn format_caps(harness: HarnessId) -> FormatCaps {
    match harness {
        HarnessId::Codex => FormatCaps {
            skill_body_max_bytes: Some(8192),
        },
        _ => FormatCaps {
            skill_body_max_bytes: None,
        },
    }
}

pub fn capabilities(harness: HarnessId, kind: ItemKind) -> KindCaps {
    use HarnessId::*;
    use ItemKind::*;
    match (harness, kind) {
        (Claude, Agent | Skill | Hook | Command) => managed(BOTH),
        (Claude, McpServer) => managed(BOTH),
        // Plugin install/remove is parked with the marketplace work.
        (Claude, Plugin) => KindCaps {
            observe: BOTH,
            toggle: BOTH,
            ..observe_only(BOTH)
        },
        (Claude, PiExtension) => unsupported(),

        (Codex, Agent | Skill | Hook) => managed(BOTH),
        // `~/.codex/prompts` is a deprecated-but-loading surface: never write.
        (Codex, Command) => observe_only(GLOBAL),
        (Codex, McpServer) => observe_only(BOTH),
        (Codex, Plugin) => observe_only(GLOBAL),
        (Codex, PiExtension) => unsupported(),

        (Opencode, Agent | Skill) => managed(BOTH),
        // Hooks render as instruction files + config refs; only those managed
        // artifacts are observable — opencode has no native hook surface.
        (Opencode, Hook) => managed(BOTH),
        (Opencode, Command) => observe_only(BOTH),
        (Opencode, McpServer) => observe_only(BOTH),
        (Opencode, Plugin) => observe_only(BOTH),
        (Opencode, PiExtension) => unsupported(),

        // Cursor is managed project-only (no global agent scope in v1), but
        // its global command/MCP surfaces exist and are observed.
        (Cursor, Agent) => managed(PROJECT),
        // Skills share the rules dir and cannot be told apart from agents.
        (Cursor, Skill) => unsupported(),
        (Cursor, Hook) => KindCaps {
            observe: BOTH,
            ..managed(PROJECT)
        },
        (Cursor, Command) => observe_only(BOTH),
        (Cursor, McpServer) => observe_only(BOTH),
        (Cursor, Plugin) => observe_only(GLOBAL),
        (Cursor, PiExtension) => unsupported(),

        (Pi, Agent | Skill) => managed(BOTH),
        // pi hooks belong to the pi-hooks extension, not to files we manage.
        (Pi, Hook) => unsupported(),
        (Pi, Command) => observe_only(BOTH),
        (Pi, McpServer) => unsupported(),
        (Pi, Plugin) => unsupported(),
        (Pi, PiExtension) => managed(BOTH),
    }
}

File: ui/src/dev/caps.ts
import type {
  CapabilityRow,
  Enforcement,
  HarnessId,
  ItemKind,
  KindCaps,
  OpSupport,
} from "@/bindings";

const PG: OpSupport = { project: true, global: true };
const P: OpSupport = { project: true, global: false };
const G: OpSupport = { project: false, global: true };
const NO: OpSupport = { project: false, global: false };

const cap = (
  observe: OpSupport,
  manage: OpSupport,
  toggle = manage,
  enforcement: Enforcement = "not-applicable",
): KindCaps => ({
  observe,
  adopt: manage,
  install: manage,
  toggle,
  remove: manage,
  refresh: manage,
  installsAs: null,
  enforcement,
});

// A hook the tool runs, versus one it only reads as instructions.
const enforcedHook = (observe: OpSupport, manage: OpSupport): KindCaps =>
  cap(observe, manage, manage, "enforced");
const advisoryHook = (observe: OpSupport, manage: OpSupport): KindCaps =>
  cap(observe, manage, manage, "advisory");

// Codex retired its prompt directory: a command is written, toggled and
// removed as a skill, while the prompts it still loads are only read.
const codexCommand: KindCaps = {
  ...cap(G, PG),
  adopt: NO,
  installsAs: "skill",
};

// Mirrors crates/core/src/harness/caps.rs closely enough for UI gating;
// the real table stays the only authority inside the app itself.
const KIND_CAPS: Record<HarnessId, Partial<Record<ItemKind, KindCaps>>> = {
  claude: {
    agent: cap(PG, PG),
    skill: cap(PG, PG),
    hook: enforcedHook(PG, PG),
    command: cap(PG, PG),
    "mcp-server": cap(PG, PG),
    plugin: cap(G, NO, G),
  },
  codex: {
    agent: cap(PG, PG),
    skill: cap(PG, PG),
    hook: enforcedHook(PG, PG),
    command: codexCommand,
    "mcp-server": cap(PG, NO),
    plugin: cap(G, NO),
  },
  opencode: {
    agent: cap(PG, PG),
    skill: cap(PG, PG),
    hook: advisoryHook(PG, PG),
    command: cap(PG, NO),
    "mcp-server": cap(PG, NO),
    plugin: cap(PG, NO),
  },
  cursor: {
    agent: cap(P, P),
    skill: cap(P, P),
    hook: advisoryHook(P, P),
    command: cap(PG, NO),
    "mcp-server": cap(PG, NO),
  },
  pi: {
    agent: cap(PG, PG),
    skill: cap(PG, PG),
    // Enforced through the pi-hooks carrier; per-item labels downgrade
    // when no settings layer Pi loads registers it.
    hook: enforcedHook(PG, PG),
    command: cap(PG, NO),
    "pi-extension": cap(PG, PG),
  },
  gemini: {
    agent: cap(PG, PG),
    skill: cap(PG, PG),
    hook: enforcedHook(PG, PG),
    command: cap(PG, PG),
    // A server is declared per scope, but whether it is on is recorded once
    // for the whole machine, so only global scope switches it.
    "mcp-server": cap(PG, PG, G),
    plugin: cap(G, NO),
  },
  copilot: {
    agent: cap(PG, PG),
    skill: cap(PG, PG),
    hook: enforcedHook(PG, PG),
    "mcp-server": cap(PG, PG),
    plugin: cap(PG, NO, PG),
  },
};

const KINDS: ItemKind[] = [
  "agent",
  "skill",
  "hook",
  "command",
  "mcp-server",
  "plugin",
  "pi-extension",
];

export function capabilityTable(): CapabilityRow[] {
  const rows: CapabilityRow[] = [];
  for (const harness of Object.keys(KIND_CAPS) as HarnessId[]) {
    for (const kind of KINDS) {
      rows.push({
        harness,
        kind,
        caps: KIND_CAPS[harness][kind] ?? cap(NO, NO),
      });
    }
  }
  return rows;
}
